clippy_lints/src/copy_refcell.rs

```diff
@@ -0,0 +1,112 @@
+use rustc_middle::ty::layout::TyAndLayout;
+use rustc_session::impl_lint_pass;
+use rustc_span::symbol::sym;
+use rustc_span::Span;
+
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::ty::implements_trait;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Detects crate-local usage of `RefCell<T>` where `T` implements `Copy`
+    ///
+    /// ### Why is this bad?
+    ///
+    /// `RefCell` has to perform extra book-keeping in order to support references, whereas `Cell` does not.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let interior_mutable = std::cell::RefCell::new(0_u8);
+    ///
+    /// *interior_mutable.borrow_mut() = 1;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let interior_mutable = std::cell::Cell::new(0_u8);
+    ///
+    /// interior_mutable.set(1);
+    /// ```
+    #[clippy::version = "1.83.0"]
+    pub COPY_REFCELL,
+    perf,
+    "usage of `RefCell<T>` where `T` implements `Copy`"
+}
+
+pub struct CopyRefcell {
+    maximum_size: u64,
+}
+
+impl CopyRefcell {
+    pub fn new(conf: &'static Conf) -> Self {
+        Self {
+            maximum_size: conf.large_cell_limit,
+        }
+    }
+
+    fn check_copy_refcell<'tcx>(&self, cx: &LateContext<'tcx>, ty: ty::Ty<'tcx>, span: Span) {
+        let Some(copy_def_id) = cx.tcx.get_diagnostic_item(sym::Copy) else {
+            return;
+        };
+
+        let ty::Adt(adt, generics) = ty.kind() else {
+            return;
+        };
+
+        if !cx.tcx.is_diagnostic_item(sym::RefCell, adt.did()) {
+            return;
+        }
+
+        let inner_ty = generics.type_at(0);
+        let Ok(TyAndLayout { layout, .. }) = cx.tcx.layout_of(cx.param_env.and(inner_ty)) else {
+            return;
+        };
+
+        if layout.size().bytes() >= self.maximum_size {
+            return;
+        }
+
+        if implements_trait(cx, inner_ty, copy_def_id, &[]) {
+            span_lint_and_help(
+                cx,
+                COPY_REFCELL,
+                span,
+                "std::cell::RefCell used with a type that implements Copy",
+                None,
+                "replace the usage of RefCell with Cell, which is simpler and zero-cost",
+            );
+        }
+    }
+}
+
+impl_lint_pass!(CopyRefcell => [COPY_REFCELL]);
+
+impl<'tcx> LateLintPass<'tcx> for CopyRefcell {
+    fn check_field_def(&mut self, cx: &LateContext<'tcx>, field_def: &'tcx FieldDef<'tcx>) {
+        // Don't want to lint macro expansions.
+        if field_def.span.from_expansion() {
+            return;
+        }
+
+        let field_ty = rustc_hir_analysis::lower_ty(cx.tcx, field_def.ty);
+        self.check_copy_refcell(cx, field_ty, field_def.ty.span);
+    }
+
+    fn check_local(&mut self, cx: &LateContext<'tcx>, local_def: &'tcx LetStmt<'tcx>) {
+        // Don't want to lint macro expansions or desugaring.
+        if local_def.span.from_expansion() || !matches!(local_def.source, LocalSource::Normal) {
```

Comment:
`from_expansion` includes desugaring

Reply:
Shall I remove the `matches!`? Or just remove the comment?

Reply:
Yeah the `matches!` is unnecessary